agent/src/protocols/coap_core/discovery_handler.rs

```diff
@@ -0,0 +1,145 @@
+    }
+}
+
+#[async_trait]
+impl DiscoveryHandler for CoAPCoREDiscoveryHandler {
+    async fn discover(&self) -> Result<Vec<DiscoveryResult>, failure::Error> {
+        let ip_addresses = &self.discovery_handler_config.ip_addresses;
+        let mut results: Vec<DiscoveryResult> = vec![];
+
+        for ip_address in ip_addresses {
+            let endpoint = format!("coap://{}:5683/well-known/core", ip_address);
+            info!("Discovering resources on endpoint {}", endpoint);
+
+            // TODO: make timeout configurable
+            let response = CoAPClient::get_with_timeout(endpoint.as_str(), Duration::from_secs(5));
+
+            match response {
+                Ok(response) => {
+                    let payload = String::from_utf8(response.message.payload)
+                        .expect("Receive payload is not a string");
+                    info!("Device responded with {}", payload);
+
+                    let mut properties: HashMap<String, String> = HashMap::new();
+                    let resources = discovery_impl::parse_link_value(payload.as_str());
+                    let resource_types: Vec<String> = resources.iter().map(|res| res.1.clone()).collect();
+
+                    properties.insert(COAP_IP_LABEL_ID.to_string(), ip_address.clone());
+                    properties.insert(COAP_RESOURCE_TYPES_LABEL_ID.to_string(), resource_types.join(","));
+
+                    for (uri, rtype) in resources {
+                        properties.insert(rtype, uri);
+                    }
+
+                    let result = DiscoveryResult::new(ip_address.as_str(), properties, false);
+
+                    results.push(result);
+                }
+                Err(e) => {
+                    info!("Error requesting resource discovery to device: {}", e);
+                }
+            }
+        }
+
+        Ok(results)
+    }
+
+    fn are_shared(&self) -> Result<bool, failure::Error> {
+        Ok(false)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    // use coap::Server;
+    // use coap_lite::{CoapRequest, CoapResponse};
+    // use std::future::Future;
+    // use std::net::SocketAddr;
+    // use std::thread::JoinHandle;
+
+    // pub fn spawn_server<
+    //     F: FnMut(CoapRequest<SocketAddr>) -> HandlerRet + Send + 'static,
+    //     HandlerRet,
+    // >(
+    //     request_handler: F,
+    // ) -> JoinHandle<()>
+    // where
+    //     HandlerRet: Future<Output = Option<CoapResponse>>,
+    // {
+    //     std::thread::Builder::new()
+    //         .name(String::from("coap server"))
+    //         .spawn(move || {
+    //             tokio::runtime::Runtime::new()
+    //                 .unwrap()
+    //                 .block_on(async move {
+    //                     let mut server = Server::new("127.0.0.1:5683").unwrap();
+
+    //                     server.run(request_handler).await.unwrap();
+    //                 })
+    //         })
+    //         .unwrap()
+    // }
```

Comment:
I'm not sure how testing could be done. Something like [httptest](https://docs.rs/httptest/0.14.0/httptest/) would be nice, but it's easier said than done for CoAP. I have tried this implementation but didn't work. The server was not up when the test tries to connect later. Maybe the server is "dropped" as soon as the function returns?

I resorted to creating a server on another terminal tab to run the tests. I'm very open to any idea.

Reply:
the most complete way that comes to mind would be to wrap `CoAPClient::get_with_timeout` in a trait and create a mock version of it.  i would then create an internal version of discover that accepts the trait as an argument.  this would allow you to test all of the code (what get_with_timeout expects as arguments, how the code in discover reacts to various return values, etc).

Reply:
`MockOnvifQueryImpl` is an example of that

Reply:
Thanks for the tip, I followed it